### PR TITLE
[doc] add GVT-g_api.rst

### PR DESCRIPTION
**Pull request by xyliu**
The document is partly generated from Linux kernel source code via
kernel-doc tool.

ACRN specific stuff are generated from ACRN Hypervisor with a few
introction.

Signed-off-by: Xinyun Liu <xinyun.liu@intel.com>

### CODE REVIEW

**Comment by NanlinXie** [comment]
Can one of the admins verify this patch?

**Comment by xyliu** [comment]
Don't merge!  . 
This is only for David to review and revise.

File: doc/api/GVT-g_api.rst [hunk 1]
```diff
@@ -0,0 +1,114 @@
+.. _GVT-g_api:
+
+===============
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: Take a look at our ACRN documentation guidelines at https://projectacrn.github.io/latest/developer-guides/doc_guidelines.html for our project conventions for writing reST files.

For the document title we just use a ``#####`` underline, so change the top of the doc to:

    .. _GVT-g_api:

    ACRN GVT-g APIs
    ###############

File: doc/api/GVT-g_api.rst [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+ACRN GVT-g APIs
+===============
+
+GVT-g is the GPU virtualization solution of Intel, it is open source and
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: Change this first sentence to:

    GVT-g is Intel's open source GPU virtualization solution and is up-streamed
    to the Linux kernel.

File: doc/api/GVT-g_api.rst [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+up-streamed to the Linux kernel. Its implementation over KVM is named KVMGT,
+over Xen is named XenGT and over ACRN is named ACRN-GT. GVT-g could exports
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: add a comma:  ``... named XenGT, and over ACRN ...

Replace the next sentence with:

    GVT-g can export multiple virtual GPU (vGPU) instances for
    virtual machine (VM) systems.

File: doc/api/GVT-g_api.rst [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+multiple virtual GPU (alias vGPU) instances for virtual machine system (a.k.a.
+VM). A VM could be assigned one instance of vGPU. The guest OS graphic driver
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: ``of a vGPU.``

File: doc/api/GVT-g_api.rst [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+need minor modification to drive the vGPU adapter in VM. Every vGPU instance
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: ``needs``
``in a VM``

    Each vGPU instance adopts the full HW GPU's acceleration capabilities
    for 3D rendering and display.

File: doc/api/GVT-g_api.rst [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+will adopts the full HW GPU’s accelerate capability for 3D render and display.
+In the following document, ACRN-GT refers to the glue layer between ACRN
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: Add a blank line before this and change to:

    In the following API documentation, ...

``between the ACRN``

File: doc/api/GVT-g_api.rst [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+hypervisor and GVT-g core device model. It works as the agency of hypervisor
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: ``It works as the agent of hypervisor-related services.``

File: doc/api/GVT-g_api.rst [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+related services. It is the very and only layer that needs to get rewritten
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: Delete ``very and``

File: doc/api/GVT-g_api.rst [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+when porting GVT-g to other specific hypervisor.  For simplicity, in the rest
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: change ``other specific`` to ``another``

File: doc/api/GVT-g_api.rst [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+of this document, GVT is used to refer to the core device model component of
+GVT-g, specifically corresponding to gvt.ko when build as module.
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: ``built as a module.``

File: doc/api/GVT-g_api.rst [hunk 11]
```diff
@@ -0,0 +1,114 @@
+.. _GVT-g_api:
+
+===============
+ACRN GVT-g APIs
+===============
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: top level headings after the title use ``*****`` for the underline, so change all your uses of ``=====`` to ``******``

File: doc/api/GVT-g_api.rst [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+will adopts the full HW GPU’s accelerate capability for 3D render and display.
+In the following document, ACRN-GT refers to the glue layer between ACRN
+hypervisor and GVT-g core device model. It works as the agency of hypervisor
+related services. It is the very and only layer that needs to get rewritten
+when porting GVT-g to other specific hypervisor.  For simplicity, in the rest
+of this document, GVT is used to refer to the core device model component of
+GVT-g, specifically corresponding to gvt.ko when build as module.
+
+Core Driver Infrastructure
+==========================
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: change to use ``******``

File: doc/api/GVT-g_api.rst [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+This section covers core driver infrastructure used by both the display
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: ``infrastructure APIs used``

File: doc/api/GVT-g_api.rst [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+and the GEM parts of the driver.
+
+Intel GVT-g Guest Support(vGPU)
+-------------------------------
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: Change all uses of ``--------`` to our convention of second level headings using ``=========``

File: doc/api/GVT-g_api.rst [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+
+.. kernel-doc:: drivers/gpu/drm/i915/i915_vgpu.c
+   :doc: Intel GVT-g guest support
+
+.. kernel-doc:: drivers/gpu/drm/i915/i915_vgpu.c
+   :internal:
+
+Intel GVT-g Host Support(vGPU device model)
+-------------------------------------------
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: change to use ``========``

File: doc/api/GVT-g_api.rst [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+.. kernel-doc:: drivers/gpu/drm/i915/intel_gvt.c
+   :doc: Intel GVT-g host support
+
+.. kernel-doc:: drivers/gpu/drm/i915/intel_gvt.c
+   :internal:
+
+
+VHM APIs called from ACRN-GVT-g 
+===============================
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: change to use ``********``

File: doc/api/GVT-g_api.rst [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+
+The Virtio and Hypervisor Service Module is a kernel module in the Service OS
+acting as a middle layer to support the device model.
+(https://projectacrn.github.io/latest/introduction/index.html#acrn-i-o-mediator).
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: Change this to be a Sphinx reference:

    (See the :ref:`ACRN-io-mediator`` introduction for details.)

File: doc/api/GVT-g_api.rst [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+VHM requires an interrupt(vIRQ) number, and exposes some API to external kernel
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: ``some APIs``

File: doc/api/GVT-g_api.rst [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+modules like GVT-g, virtIO BE service running in kernel space. It exposes a
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: change to ``modules such as GVT-g and the virtIO BE service running in kernel space.``

change ``It exposes`` to ``VHM exposes``

File: doc/api/GVT-g_api.rst [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+char device node in user space, and only interacts with DM, and DM routes IO
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: change to:

    and only interacts with the DM.  The DM routes I/O requests and responses
    to and from other modules via this char device to and from the VHM.

File: doc/api/GVT-g_api.rst [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+request/response from/to other modules to/from VHM via the char device. DM may
+use VHM for hypervisor service (including remote memory map), and VHM may
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: make this two sentences by ending:

    remote memory map).  VHM may directly ...

File: doc/api/GVT-g_api.rst [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+directly service the request such as for the remote memory map, or invoke
+hypercall, also sends IO response to user space modules, notified by vIRQ
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: start another sentence here too:

   or invoke hypercall.  VMH also sends I/O responses to user space modules,
   notified by vIRQ injections.

File: doc/api/GVT-g_api.rst [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+
+.. kernel-doc:: include/linux/vhm/acrn_vhm_mm.h 
+   :functions: unset_mmio_map
+
+.. kernel-doc:: include/linux/vhm/acrn_vhm_mm.h 
+   :functions: update_memmap_attr
+
+
+ACRN-GT MPT interface
+=====================
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: change to use ``******``

File: doc/api/GVT-g_api.rst [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+
+.. kernel-doc:: drivers/gpu/drm/i915/gvt/acrn-gvt.c
+
+.. kernel-doc:: drivers/gpu/drm/i915/gvt/hypercall.h
+
+.. kernel-doc:: drivers/gpu/drm/i915/gvt/mpt.h
+
+GVT-g intel_gvt_ops interface
+=============================
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: change the rest of these to use ``******``

File: doc/api/GVT-g_api.rst (hunk 1 above)
Comment (xinyunliu) [reply doc/api/GVT-g_api.rst]: Thanks.  Will convert to ACRN style instead of kernel-doc style.

File: doc/api/GVT-g_api.rst (hunk 3 above)
Comment (gvancuts) [reply doc/api/GVT-g_api.rst]: We should also add a couple of  subject ('it), i.e.:
...over Xen it is named XenGT, and over ACRN it is named ACRN-GT

A different question, it's the first time I see the term ACRN-GT coined, any reason why we're not using ACRNGT (all attached, just like for Xen and KVM)? (I personnally like it better with a hyphen but it's a divergence from the other implementations)

File: doc/api/GVT-g_api.rst [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+-------------------------------------------
+
+.. kernel-doc:: drivers/gpu/drm/i915/intel_gvt.c
+   :doc: Intel GVT-g host support
+
+.. kernel-doc:: drivers/gpu/drm/i915/intel_gvt.c
+   :internal:
+
+
+VHM APIs called from ACRN-GVT-g
```
Comment (gvancuts) [inline doc/api/GVT-g_api.rst]: Can you spell out the `VHM` acronym (first time it appear). I believe it stands for `Virtio and Hypervisor Service Module`, doesn't it?

You use the term `ACRN-GT` throughout the doc so unless this is something slightly different, we should stick to that, i.e.

File: doc/api/GVT-g_api.rst [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+   :doc: Intel GVT-g host support
+
+.. kernel-doc:: drivers/gpu/drm/i915/intel_gvt.c
+   :internal:
+
+
+VHM APIs called from ACRN-GVT-g 
+===============================
+
+The Virtio and Hypervisor Service Module is a kernel module in the Service OS
```
Comment (gvancuts) [inline doc/api/GVT-g_api.rst]: Add `(VHM)` after `Virtio and Hypervisor Service Module` to make clear these are the same thing.

File: doc/api/GVT-g_api.rst (hunk 18 above)
Comment (gvancuts) [inline doc/api/GVT-g_api.rst]: Update "interrupt(vIRQ)" to "interrupt (``vIRQ``)" (add space and double-backticks)

File: doc/api/GVT-g_api.rst (hunk 20 above)
Comment (gvancuts) [reply doc/api/GVT-g_api.rst]: Can you put char in double-backticks (``char``)

File: doc/api/GVT-g_api.rst (hunk 19 above)
Comment (gvancuts) [inline doc/api/GVT-g_api.rst]: I've always seen `virtio` spelled in lower case so unless there is a reason to deviate from this, let's stick to `virtio`.

Can you spell out BE, i.e. `virtio backend (BE)`

File: doc/api/GVT-g_api.rst (hunk 23 above)
Comment (gvancuts) [reply doc/api/GVT-g_api.rst]: Can you spell out what MPT stand for?

File: doc/api/GVT-g_api.rst [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+
+GVT-g intel_gvt_ops interface
+=============================
+
+
+GVT-g APIs invoked by Intel graphics driver
+===========================================
+
+
+ACRN-GT Sysfs interface
```
Comment (gvancuts) [inline doc/api/GVT-g_api.rst]: Use lower-case sysfs and enclose it in double-backticks: ``sysfs``

File: doc/api/GVT-g_api.rst [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+and the GEM parts of the driver.
+
+Intel GVT-g Guest Support(vGPU)
+-------------------------------
+
+.. kernel-doc:: drivers/gpu/drm/i915/i915_vgpu.c
+   :doc: Intel GVT-g guest support
+
+.. kernel-doc:: drivers/gpu/drm/i915/i915_vgpu.c
+   :internal:
```
[the remark this answers is not in the document]
Comment (xinyunliu) [reply doc/api/GVT-g_api.rst]: This section is from kernel. The final rendering result is here:
https://01.org/linuxgraphics/gfx-docs/drm/gpu/i915.html#intel-gvt-g-guest-support-vgpu

Now I am thinking about to split the patch to 3 small one:
   1. Skeleton
   2. Pure API document (extract from code)
   3.  More descriptions, mostly from the HLD documents.

And Item 3, can be improved step by step.

File: doc/api/GVT-g_api.rst (hunk 4 above)
Comment (xinyunliu) [reply doc/api/GVT-g_api.rst]: Will change it to "on vGPU instance".   
In current code, GVT-g allows 8 vGPUs at most. And each VM has one vGPU instance.

File: doc/api/GVT-g_api.rst [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+ACRN GVT-g APIs
+===============
+
+GVT-g is the GPU virtualization solution of Intel, it is open source and
+up-streamed to the Linux kernel. Its implementation over KVM is named KVMGT,
+over Xen is named XenGT and over ACRN is named ACRN-GT. GVT-g could exports
+multiple virtual GPU (alias vGPU) instances for virtual machine system (a.k.a.
+VM). A VM could be assigned one instance of vGPU. The guest OS graphic driver
+need minor modification to drive the vGPU adapter in VM. Every vGPU instance
+will adopts the full HW GPU’s accelerate capability for 3D render and display.
```
Comment (xinyunliu) [inline doc/api/GVT-g_api.rst]: s/adopts/adopt

File: doc/api/GVT-g_api.rst (hunk 19 above)
Comment (xinyunliu) [reply doc/api/GVT-g_api.rst]: OK. followed https://projectacrn.github.io/latest/developer-guides/primer.html#virtio-devices,  will changed to:

Virtio back-end(BE)

File: doc/api/GVT-g_api.rst [hunk 30]
```diff
@@ -0,0 +1,298 @@
...
+
+sysfs nodes
+===========
+
+In below examples all accesses to these interfaces are via bash command
+``echo`` or ``cat``. This is a quick and easy way to get/control things. But
+when these operations fails, it is impossible to get respective error code by
+this way.
+
+When accessing sysfs entries, people should use library functions like
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: change ``like`` to ``such as``

``like`` should be used for comparison (it floats like a butterfly) while ``such as`` is used for examples (I prefer colors such as red and blue).

File: doc/api/GVT-g_api.rst [hunk 31]
```diff
@@ -0,0 +1,298 @@
...
+   :doc: Intel GVT-g host support
+
+.. kernel-doc:: drivers/gpu/drm/i915/intel_gvt.c
+   :internal:
+
+
+VHM APIs called from ACRN-GT
+****************************
+
+The Virtio and Hypervisor Service Module( ``VHM`` ) is a kernel module in the
```
[the remark this answers is not in the document]
Comment (xyliu) [reply doc/api/GVT-g_api.rst]: Thanks, will update it.

File: doc/api/GVT-g_api.rst [hunk 32]
```diff
@@ -0,0 +1,298 @@
...
+Service OS acting as a middle layer to support the device model.(See the
```
[the remark this answers is not in the document]
Comment (xyliu) [reply doc/api/GVT-g_api.rst]: Thanks, will update it.

File: doc/api/GVT-g_api.rst (hunk 30 above)
Comment (xyliu) [reply doc/api/GVT-g_api.rst]: updated.

File: doc/api/GVT-g_api.rst [hunk 33]
```diff
@@ -0,0 +1,297 @@
...
+
+
+VHM APIs called from ACRN-GT
+****************************
+
+The Virtio and Hypervisor Service Module (VHM) is a kernel module in the
+Service OS acting as a middle layer to support the device model. (See the
+:ref:`ACRN-io-mediator` introduction for details.)
+
+VHM requires an interrupt( ``vIRQ`` ) number, and exposes some APIs to external
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: You missed some of these in your recent update:

    an interrupt (vIRQ) number

File: doc/api/GVT-g_api.rst [hunk 34]
```diff
@@ -0,0 +1,297 @@
...
+kernel modules such as GVT-g and the Virtio back-end( ``BE`` ) service running
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: Virtio back-end (BE) service

Comment (xinyunliu) [reply doc/api/GVT-g_api.rst]: Yes.  Thanks!  Updated :)

**Comment by dbkinder** [comment]
@jren1 @xyliu @hwang37 Do we want to merge this in time to catch the 0.1 release tag?

**Comment by hwang37** [comment]
If we can merge this PR now, then we can catch up v0.1 tag window, that'll be great. But it's also fine to merge later if we can't make it due to time constraint.

**Comment by dbkinder** [comment]
I had a chance to actually run the doc build with these changes and I'm seeing lots of warnings that need to be addressed before merging this PR:

```
../../acrn-kernel/include/linux/vhm/vhm_vm_mngt.h:58: warning: Cannot understand  * @file vhm_vm_mngt.h
 on line 58 - I thought it was a doc line
../../acrn-kernel/include/linux/vhm/vhm_vm_mngt.h:58: warning: Cannot understand  * @file vhm_vm_mngt.h
 on line 58 - I thought it was a doc line
../../acrn-kernel/include/linux/vhm/vhm_vm_mngt.h:58: warning: Cannot understand  * @file vhm_vm_mngt.h
 on line 58 - I thought it was a doc line
../../acrn-kernel/include/linux/vhm/vhm_vm_mngt.h:58: warning: Cannot understand  * @file vhm_vm_mngt.h
 on line 58 - I thought it was a doc line
../../acrn-kernel/include/linux/vhm/acrn_vhm_ioreq.h:56: warning: Cannot understand  * @file acrn_vhm_ioreq.h
 on line 56 - I thought it was a doc line
../../acrn-kernel/include/linux/vhm/acrn_vhm_mm.h:55: warning: Cannot understand  * @file acrn_vhm_mm.h
 on line 55 - I thought it was a doc line
../../acrn-kernel/include/linux/vhm/acrn_vhm_mm.h:55: warning: Cannot understand  * @file acrn_vhm_mm.h
 on line 55 - I thought it was a doc line
../../acrn-kernel/include/linux/vhm/acrn_vhm_mm.h:55: warning: Cannot understand  * @file acrn_vhm_mm.h
 on line 55 - I thought it was a doc line
../../acrn-kernel/include/linux/vhm/acrn_vhm_mm.h:55: warning: Cannot understand  * @file acrn_vhm_mm.h
 on line 55 - I thought it was a doc line
../../acrn-kernel/include/linux/vhm/acrn_vhm_mm.h:55: warning: Cannot understand  * @file acrn_vhm_mm.h
 on line 55 - I thought it was a doc line
../../acrn-kernel/include/linux/vhm/acrn_vhm_mm.h:55: warning: Cannot understand  * @file acrn_vhm_mm.h
 on line 55 - I thought it was a doc line
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mpt.h:54: warning: No description found for parameter 'dev'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mpt.h:54: warning: No description found for parameter 'gvt'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mpt.h:54: warning: No description found for parameter 'ops'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mpt.h:67: warning: No description found for parameter 'dev'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mpt.h:67: warning: No description found for parameter 'gvt'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mpt.h:83: warning: No description found for parameter 'vgpu'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mpt.h:99: warning: No description found for parameter 'vgpu'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mpt.h:119: warning: No description found for parameter 'vgpu'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mpt.h:247: warning: No description found for parameter 'gfn'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mpt.h:247: warning: Excess function parameter 'gpfn' description in 'intel_gvt_hypervisor_gfn_to_mfn'

../../acrn-kernel/drivers/gpu/drm/i915/gvt/cfg_space.c:66: warning: No description found for parameter 'vgpu'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/cfg_space.c:66: warning: No description found for parameter 'off'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/cfg_space.c:66: warning: No description found for parameter 'src'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/cfg_space.c:66: warning: No description found for parameter 'bytes'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/cfg_space.c:100: warning: No description found for parameter 'vgpu'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/cfg_space.c:100: warning: No description found for parameter 'offset'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/cfg_space.c:100: warning: No description found for parameter 'p_data'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/cfg_space.c:100: warning: No description found for parameter 'bytes'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/cfg_space.c:272: warning: No description found for parameter 'vgpu'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/cfg_space.c:272: warning: No description found for parameter 'offset'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/cfg_space.c:272: warning: No description found for parameter 'p_data'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/cfg_space.c:272: warning: No description found for parameter 'bytes'

../../acrn-kernel/drivers/gpu/drm/i915/gvt/mmio.c:47: warning: No description found for parameter 'gpa'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mmio.c:309: warning: No description found for parameter 'dmlr'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mmio.c:1: warning: no structured comments found
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mmio.c:47: warning: No description found for parameter 'gpa'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mmio.c:309: warning: No description found for parameter 'dmlr'
../../acrn-kernel/drivers/gpu/drm/i915/gvt/mmio.c:1: warning: no structured comments found
```

**Comment by xinyunliu** [comment]
@dbkinder, warnings are from kernel source code and not possible to let this patch to fix.

For the doxygen format issue in VHM module, they doesn't block doc generation.  For gvt module, the patch is submitted to acre-dev mailing list for review . @minhe1 @yakuizhao0 @dbkinder https://github.com/projectacrn/acrn-kernel/pull/28 this is kernel part changes, I did't generate it because it's not reviewed yet in mailing list.

**Comment by dbkinder** [comment]
The document build is expected to have no error or warning messages reported.  We have a way to document "known" messages if they're out of our control to fix.  Perhaps you can submit a PR to the i915 driver code to get the kernel-doc comments fixed?  As for the VHM module, we can control that and should figure out the right documentation method to not produce errors.  (You can't use a kernel-doc directive to get API info from those modules, you need to use the sphinx/breathe directive instead.)

**Comment by hwang37** [comment]
Agree with David. VHM should follow kernel-doc tool format after this v0.1.
Will ask Chaohong to follow.

**Comment by xinyunliu** [comment]
@dbkinder, the dependency in [acrn-kernel](https://github.com/projectacrn/acrn-kernel/pull/28) is merged and could you please try again to see if any warning there? Thanks!

File: doc/api/GVT-g_api.rst [hunk 35]
```diff
@@ -0,0 +1,297 @@
...
+kernel modules such as GVT-g and the Virtio back-end (BE) service running in
+kernel space.  VHM exposes a ``char`` device node in user space, and only
+interacts with DM. The DM routes I/O request and response from and to other
+modules via the ``char`` device to and from VHM. DM may use VHM for hypervisor
+service (including remote memory map). VHM may directly service the request
+such as for the remote memory map, or invoke hypercall. VHM also sends I/O
+responses to user space modules, notified by vIRQ injections.
+
+
+.. kernel-doc:: include/linux/vhm/vhm_vm_mngt.h
```
Comment (dbkinder) [inline doc/api/GVT-g_api.rst]: If you didn't know, you can list more than one function (separated by a space) for this kernel-doc directive:

    .. kernel-doc:  include/linux/vhm/vhm_vm_mngt.h
       :functions: put-vm vhm_get_vm_info vhm_inject_msi vmh_vm_gpa2hpa

Every ``.. kernel-doc::`` directive use makes Sphinx scan the source file, so if you combine the function names for a given source file, Sphinx will only have to open and scan that file one time.

**Comment by dbkinder** [comment]
@xinyunliu let me know if you think this is ready to merge or do we want further technical review on the content?

File: doc/api/GVT-g_api.rst (hunk 35 above)
Comment (xinyunliu) [reply doc/api/GVT-g_api.rst]: Thanks for introduce this great method. It's better if it accepts the "\" to separate to several lines.

**Comment by xinyunliu** [comment]
@dbkinder, I have updated the doc per your  advise, now it looks more clean.  If could, please approve and merge, thanks a lot!

**Comment by dbkinder** [comment]
@xinyunliu Can I assume the CI failure is not related to this documentation change?

**Comment by xinyunliu** [comment]
@dbkinder @NanlinXie @lijinxia , the  PR should not affect the ACRN system test, this is document change only.   Can you verify it?

But when I verify the patch in docker environment, there is no breathe installed. So I got error like below.  It's resolved by follow doc/tutorials/docbuild.rst, 

``` bash
   cd ~/projectacrn/acrn-hypervisor/doc
   pip3 install --user -r scripts/requirements.txt
```

> $make html
> make: [Makefile:52: html] Error 2 (ignored)
> 
> New errors/warnings found, please fix them:
> ==============================================
> 
> 
> Extension error:
> Could not import extension breathe (exception: No module named 'breathe')

**Comment by dbkinder** [comment]
I can generate the docs cleanly in my local environment, so I'm going to merge this despite the check failure.

**Pull request merged**